### ISSUE TITLE
Override Bootstrap classes in Bootstrap widgets

### ISSUE THREAD

**Issue by alexandernst** [yiisoft/yii2#8229]
I have a NavBar and I want to override all the classes that Yii is putting on that widget. After searching inn Google and reading the docs, I found that this code:

```
Yii::$container->set('yii\bootstrap\NavBar', [
    'containerOptions' => [
        'class' => ''
    ]
]);

NavBar::begin([
    'containerOptions' => [
        'class' => 'no-horizontal-padding navbar-content'
    ]
]);
```

is supposed to override (delete, actually) all classes in the container of my NavBar, but this is not the case. I keep seeing my 2 classes and the default NavBar classes.

How can I remove those classes and keep only mine?

Please refrain any hacky solutions like removing the classes with JS on the client side. I'm looking for a proper way of doing this.

```
<nav id="w0" class="navbar-inverse navbar" role="navigation">
    <div class="container-fluid no-horizontal-padding">
        <div class="navbar-header">
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#w0-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
        </div>
        <div id="w0-collapse" class="no-horizontal-padding navbar-content collapse navbar-collapse">
        </div> <!-- The classes that should be  removed are here -->
    </div>
</nav>
```

**Comment by cebe** [comment]
the reason for this behavior is this code: https://github.com/yiisoft/yii2-bootstrap/blob/master/NavBar.php#L97
It assumes that there is no use case where you would not want the class navbar on the element. Why do you want to remove it?

**Comment by alexandernst** [comment]
@cebe Because it's setting a class that it's setting a `display: block;`, but I want a `display: flex;`.
 I tried using `!important` in my class, but it's not working because (I'm guessing) bootstrap classes take precedence. Anyways, even if it did worked, I wouldn't be happy about using `!important` all over my code just because I can't remove the classes Yii is imposing on widgets.

**Comment by alexandernst** [comment]
Just to make my point clear: I'm not asking to remove all classes, but just the classes from the `containerOptions` part of the NavBar, that is, the `<div id="w0-collapse"` div in my code.

**Comment by alexandernst** [comment]
@cebe I could work on this issue as I'm interested in it. How should I implement this? The same way as `\Yii::$container->set(class, array())` works?

**Comment by cebe** [comment]
I am not sure I really understood the issue yet.

```php
        NavBar::begin([
            'containerOptions' => [
                'class' => 'no-horizontal-padding navbar-content'
            ]
        ]);
```

prints the following HTML:

```html
<nav id="w0" class="navbar navbar-default" role="navigation"><div class="container"><div class="navbar-header"><button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#w0-collapse"><span class="sr-only">Toggle navigation</span>
<span class="icon-bar"></span>
<span class="icon-bar"></span>
<span class="icon-bar"></span></button></div><div id="w0-collapse" class="no-horizontal-padding navbar-content collapse navbar-collapse">
```

The classes `collapse navbar-collapse` are added because this ensures the navbar gets collapsed on small screens. Without these your navbar stays open (all menu items visible) on a small screen.

I think it is better to solve this with css. The following rule should work in your case:

```css
.navbar-collapse.collapse {
    display: flex !important;
}
```

**Comment by alexandernst** [comment]
@cebe Sorry for the late reply, I had some technical difficulties.
There are 2 main problems here.

The first one is that I actually don't want those classes on my HTML code because, as I said, bootstrap is somehow (can't say why) taking precedence over my classes and the rules in my classes aren't working (not even with `!important`). Go and try, you'll see that the `display` rule actually isn't applied because bootstrap has it's own rule (which also has an `!important` flag).

The second one: Even if `!important` worked (which it doesn't), I wouldn't like to be forced to use `!important` all over my  CSS code just because I can't remove some classes that I don't need anyways.

**Comment by alexandernst** [comment]
@cebe Hi! Did you have time to test this?

**Comment by alexandernst** [comment]
@cebe If you want me, I can setup a simple demo in a repo so you can clone it and see what I'm talking about.

**Comment by cebe** [comment]
please do, in the example i tried with the basic app, the `!important` override worked just fine.

**Comment by alexandernst** [comment]
@cebe Done! https://github.com/alexandernst/tmp_yii_8229

You'll see this is just a basic app template. The only thing I added is the `testclass` to the navbar (check `site.css` and the layout).

I set the `testclass`to `display: flex !important`, but it gets overriden by Bootstrap's classes.

**Comment by alexandernst** [comment]
@cebe Could you see the bug I was talking about? Do you know what is causing it?

**Comment by alexandernst** [comment]
Closing as @klimov-paul just merged #8458 :)

**Comment by klimov-paul** [comment]
Resolved by https://github.com/yiisoft/yii2-bootstrap/pull/52
